core/shared/src/main/scala/zio/Has.scala

```diff
@@ -217,8 +220,59 @@ object Has {
      */
     def update[B: Tag](f: B => B)(implicit ev: Self MustHave B): Self =
       self.add(f(get[B]))
+
+    /**
+     * Adds a service to the environment.
+     */
+    def addAt[K, A](k: K, a: A)(implicit ev: Self <:< HasMany[K, A], tagged: Tag[Map[K, A]]): Self = {
+      val tag = taggedTagType(tagged)
+
+      val map = self.map
+        .getOrElse(
+          tag,
+          self.cache.getOrElse(
+            tag,
+            Map.empty
+          )
+        )
+        .asInstanceOf[Map[K, A]]
+        .updated(k, a)
+
+      new Has(self.map + (tag -> map)).asInstanceOf[Self]
+    }
+
+    /**
+     * Retrieves a service at the specified key from the environment.
+     */
+    def getAt[K, A](k: K)(implicit ev: Self <:< HasMany[K, A], tagged: Tag[Map[K, A]]): Option[A] = {
```

Comment:
Any thoughts on `Tagged[Map[K, A]]` versus `Tagged[K], Tagged[A]`.

Reply:
Yes. One should be derivable from the other but probably better to require the most primitive components.

Reply:
Scratch that. We actually need to require the `Tagged[Map[K, A]]` to work around an issue with the tag machinery in Scala 3. Opened zio/izumi-reflect#208 for it.